src/compiler.rkt

```diff
@@ -284,356 +283,294 @@
 
     ; This case is weird. if we have a cancellation in fma -> ival-mult in fma should be in higher precision
     (when (equal? op ival-fma)
-      (set! final-parent-precision (+ final-parent-precision new-exponents)))
+      (set! final-parent-precision (+ final-parent-precision (first new-exponents))))
     
     (when (equal? final-parent-precision (*max-mpfr-prec*))         ; Early stopping
       (*sampling-iteration* (*max-sampling-iterations*)))
     (vector-set! vprecs-new (- n varc) final-parent-precision)
-    
-    (define child-exponents (+ exps-from-above new-exponents))
-    (for ([x (in-list tail-registers)])
-      (when (>= x varc)   ; when tail register is not a variable
-        (vector-set! vprecs-new (- x varc)
-                     (max ; check whether this op already has a precision that is higher
-                      (vector-ref vprecs-new (- x varc))
-                      child-exponents))))))
 
-; Function calculates an exponent for a certain output and input using condition formulas,
+    (for ([x (in-list tail-registers)]
+          [new-exp (in-list new-exponents)]
+          #:when (>= x varc)) ; when tail register is not a variable
+      (when (> (+ exps-from-above new-exp (vector-ref vstart-precs (- x varc))) (vector-ref vprecs-new (- x varc))) ; check whether this op already has a precision that is higher
```

Comment:
Made an additional check, because `vprecs-new` already contains final precisions when `exps-from-above + new-exp` is just an exponent without adding base precision. Adding base precision and comparing is a more correct way, otherwise as it was before it can lose ~100bits of precision